test/e2e/framework/framework.go

```diff
@@ -44,6 +44,7 @@ import (
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/restmapper"
 	scaleclient "k8s.io/client-go/scale"
+	e2ekubectl "k8s.io/kubernetes/test/e2e/framework/kubectl"
```

Comment:
This makes an invalid dependency that the core framework depends on sub framework packages.

Reply:
Hmm, @oomichi what is the alternative if I'm taking `KubectlCmd` out of `utils.go` but `KubectlCmd` is still needed within `framework.go`?

Reply:
Here's the original ticket: https://github.com/kubernetes/kubernetes/issues/77095

Where the proposed structure was `test/e2e/framework/kubectl`. `framework.go` needs kubectl functions in order to do some of the things it does. So I'd have to import that dependency in order to use it.

@oomichi do you propose that I add this `kubectl_utils` file to the already existing `test/e2e/kubectl` package? And then import the dependency from there like:
```
e2ekubectl "k8s.io/kubernetes/test/e2e/kubectl"
```
in order to avoid importing a framework sub-package within framework?

Also, do we consider all of those sub-package dependencies in this file, `test/e2e/framework/framework.go`, invalid? If so, maybe we need to make a ticket to update that as well.

Reply:
IIUC, if needing this dependency for this separation, we should not separate this from the core framework to a subpackage.
Otherwise, we need to revert this for fixing the invalid dependency like https://github.com/kubernetes/kubernetes/pull/84509

Reply:
So, if `framework.go` needs it as a dependency(`KubectlCmd` in this case), we shouldn't separate it? That's what we mean by `invalid dependency`? @oomichi

Reply:
I see an interesting way to solve both of the issues for this PR:
The reason why `e2ekubectl` is needed within this file in the framework is because it is used in [kubectlExec()](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go#L662) which in turn is used by [kubectlExecWithRetry()](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go#L635).

`kubectlExecWithRetry()` is used by the following methods within the framework:
* [`WriteFileViaContainer()`](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go#L473)
* [`ReadFileViaContainer()`](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go#L490)
* [`CheckFileSizeViaContainer()`](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go#L501)

These methods are used by e2e tests
```
$ grep -nr "WriteFileViaContainer"
e2e/storage/pd.go:185:					framework.ExpectNoError(f.WriteFileViaContainer(host0Pod.Name, containerName, testFile, testFileContents))
e2e/storage/pd.go:292:						framework.ExpectNoError(f.WriteFileViaContainer(host0Pod.Name, containerName, testFile, testFileContents))
e2e/storage/pd.go:386:				framework.ExpectNoError(f.WriteFileViaContainer(host0Pod.Name, containerName, testFile, testFileContents))
e2e/framework/framework.go:472:// WriteFileViaContainer writes a file using kubectl exec echo <contents> > <path> via specified container
e2e/framework/framework.go:474:func (f *Framework) WriteFileViaContainer(podName, containerName string, path string, contents string) error {
e2e/node/security_context.go:215:	err = f.WriteFileViaContainer(pod.Name, pod.Spec.Containers[0].Name, testFilePath, testContent)
```

```
$ grep -nr "ReadFileViaContainer"
e2e/storage/pd.go:459:		v, err := f.ReadFileViaContainer(podName, containerName, filePath)
e2e/framework/framework.go:490:// ReadFileViaContainer reads a file using kubectl exec cat <path>.
e2e/framework/framework.go:491:func (f *Framework) ReadFileViaContainer(podName, containerName string, path string) (string, error) {
e2e/auth/service_accounts.go:227:		mountedToken, err := f.ReadFileViaContainer(pod.Name, pod.Spec.Containers[0].Name, path.Join(serviceaccount.DefaultAPITokenMountPath, v1.ServiceAccountTokenKey))
e2e/auth/service_accounts.go:229:		mountedCA, err := f.ReadFileViaContainer(pod.Name, pod.Spec.Containers[0].Name, path.Join(serviceaccount.DefaultAPITokenMountPath, v1.ServiceAccountRootCAKey))
e2e/auth/service_accounts.go:231:		mountedNamespace, err := f.ReadFileViaContainer(pod.Name, pod.Spec.Containers[0].Name, path.Join(serviceaccount.DefaultAPITokenMountPath, v1.ServiceAccountNamespaceKey))
e2e/node/security_context.go:217:	content, err := f.ReadFileViaContainer(pod.Name, pod.Spec.Containers[0].Name, testFilePath)
e2e/node/security_context.go:269:	content, err = f.ReadFileViaContainer(pod.Name, "test-container", testFilePath)
```

```
$ grep -nr "CheckFileSizeViaContainer"
e2e/framework/framework.go:501:// CheckFileSizeViaContainer returns the list of file size under the specified path.
e2e/framework/framework.go:502:func (f *Framework) CheckFileSizeViaContainer(podName, containerName, path string) (string, error) {
```

So the last one it seems we can just delete, the other two merit being moved within `e2ekubectl`.

Reply:
@MorrisLaw 
I should explain `invalid dependency` means early, sorry about that.
We defined `e2e core framework` and `subpackages of e2e framework` and `e2e core framework` consists of `test/e2e/framework/*.go` files and `subpackages of e2e framework` consist of `test/e2e/framework/*/*.go` files. In this case, `est/e2e/framework/kubectl` is a subpackage.
The valid dependency is that subpackages depend on `e2e core framework` because `e2e core framework` functionality is minimum set of testing framework and subpackages can use the functionality. But if adding opposite dependency like this PR, circular dependency happens between `e2e core framework` and `subpackages of e2e framework`.
We are trying to solve this invalid dependency at multiple issues like https://github.com/kubernetes/kubernetes/issues/84380 https://github.com/kubernetes/kubernetes/issues/81427 https://github.com/kubernetes/kubernetes/issues/81245 https://github.com/kubernetes/kubernetes/issues/81232

Reply:
@alejandrox1 
That is a nice point!
At least, we need to remove the last function `CheckFileSizeViaContainer` before doing this.
I feel it is better to clean up functions in `e2e core framework` like https://github.com/kubernetes/kubernetes/issues/84380 before separations from `e2e core framework`.

Reply:
Thank you for the thorough description @oomichi !

Yes, that circular dependency issue was annoying. I didn't realize how common of an issue this was... I'm glad you pointed out those tickets to me.

For `kubectl` related functions coming out of `util`, @alejandrox1 and I came up with a way to avoid the need to import `TestContext`(which would require the circular dependency of importing `framework` within the new `kubectl_util` file) by creating this `TestKubeconfig` struct that only contains the fields we need to run `KubectlCmd`.

That way, within `framework.go` when we call `KubectlCmd` we can first create a `TestKubeconfig` with `NewTestKubeconfig`. And since we call `KubectlCmd` on `TestKubeconfig` I can get access to the fields like `Host`, `CertDir` coming from `TestContext` without needing to deal with that inside of `kubectl_utils.go`.

Reply:
But, since @alejandrox1  pointed out those particular functions that depend on it. I can remove the need to import `e2ekubectl ...` in `framework` at all if I abstract out those particular functions (: